File: crud_java/src/main/java/com/example/crud_java/presentation/global/GlobalExceptionHandler.java
package com.example.crud_java.presentation.global;

import com.example.crud_java.application.common.exception.BusinessException;
import com.fasterxml.jackson.core.JsonParseException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import static org.springframework.http.HttpStatus.BAD_REQUEST;
import static org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {
    @ExceptionHandler(BusinessException.class)
    protected ResponseEntity<ExceptionResponse> handleBusinessException(BusinessException e) {
        log.error("BusinessException", e);
        ExceptionResponse response = ExceptionResponse.of(e.getMessage());
        return ResponseEntity.status(BAD_REQUEST).body(response);
    }

    @ExceptionHandler(JsonParseException.class)
    protected ResponseEntity<ExceptionResponse> handleJsonParseException(JsonParseException e) {
        log.error("JsonParseException", e);
        ExceptionResponse response = ExceptionResponse.of(e.getMessage());
        return ResponseEntity.status(BAD_REQUEST).body(response);
    }

    @ExceptionHandler(Exception.class)
    protected ResponseEntity<ExceptionResponse> handleException(Exception e) {
        log.error("Exception", e);
        ExceptionResponse response = ExceptionResponse.of(e.getMessage());
        return ResponseEntity.status(INTERNAL_SERVER_ERROR).body(response);
    }
}

File: crud_kotlin/src/main/kotlin/com/example/crud_kotlin/presentation/global/GlobalExceptionHandler.kt
package com.example.crud_kotlin.presentation.global

import com.example.crud_kotlin.application.common.exception.BusinessException
import com.example.crud_kotlin.application.common.log.Log
import com.fasterxml.jackson.core.JsonParseException
import org.springframework.http.HttpStatus.*
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler : Log {

    @ExceptionHandler(BusinessException::class)
    protected fun handleBusinessException(e: BusinessException): ResponseEntity<ExceptionResponse> {
        log.error("BusinessException", e)
        return ResponseEntity.status(BAD_REQUEST).body(
            ExceptionResponse(e.message)
        )
    }

    @ExceptionHandler(JsonParseException::class)
    protected fun handleJsonParseException(e: JsonParseException): ResponseEntity<ExceptionResponse> {
        log.error("JsonParseException", e)
        return ResponseEntity.status(BAD_REQUEST).body(
            ExceptionResponse(e.message)
        )
    }

    @ExceptionHandler(Exception::class)
    protected fun handleJsonParseException(e: Exception): ResponseEntity<ExceptionResponse> {
        log.error("Exception", e)
        return ResponseEntity.status(INTERNAL_SERVER_ERROR).body(
            ExceptionResponse(e.message)
        )
    }
}